exchanges/orderbook/linked_list.go

```diff
@@ -0,0 +1,618 @@
+			}
+
+			if tip.next == nil {
+				if shiftBookmark(tip, &bookmark, &ll.head, updts[x]) {
+					continue updates
+				}
+			}
+		}
+		n := stack.Pop()
+		n.value = updts[x]
+		if insertNodeAtBookmark(bookmark, n) {
+			ll.head = n
+		}
+		ll.length++
+	}
+	return nil
+}
+
+// insertUpdates inserts new updates for asks based on price level
+func (ll *asks) insertUpdates(updts Items, stack *stack) error {
+	for x := range updts {
+		var prev *node
+		for tip := &ll.head; ; tip = &(*tip).next {
+			if *tip == nil { // Head is empty
+				insertHeadSpecific(tip, updts[x], stack)
+				ll.length++
+				break // Continue updates
+			}
+
+			if (*tip).value.Price == updts[x].Price { // Price already found
+				return fmt.Errorf("%w for price %f",
+					errCollisionDetected,
+					updts[x].Price)
+			}
+
+			// Correct position/alignment found for price level
+			if (*tip).value.Price > updts[x].Price {
+				n := stack.Pop()
+				n.value = updts[x]
+				n.prev = prev
+				ll.length++
+				if prev == nil {
+					// Place new node in front of current node
+					(*tip).prev = n
+					n.next = *tip
+					// Replace head entry
+					*tip = n
+				} else {
+					old := *tip
+					prev.next = n
+					n.next = old
+					old.prev = n
+				}
+				break // Continue updates
+			}
+
+			if (*tip).next == nil { // Tail
+				insertAtTail(&ll.linkedList, tip, updts[x], stack)
+				break // Continue updates
+			}
+			prev = *tip
+		}
+	}
+	return nil
+}
+
+// move moves a node from a point in a node chain to another node position,
+// this left justified towards head as element zero is the top of the depth
+// side. (can inline)
+func move(head **node, from, to *node) {
+	if from.next != nil { // From is at tail
+		from.next.prev = from.prev
+	}
+	if from.prev == nil { // From is at head
+		(*head).next.prev = nil
+		*head = (*head).next
+	} else {
+		from.prev.next = from.next
+	}
+	// insert from node next to 'to' node
+	if to.prev == nil { // Destination is at head position
+		*head = from
+	} else {
+		to.prev.next = from
+	}
+	from.prev = to.prev
+	to.prev = from
+	from.next = to
+}
+
+// deleteAtTip removes a node from tip target returns old node (can inline)
+func deleteAtTip(ll *linkedList, tip **node) *node {
+	// Old is a placeholder for current tips node value to push
+	// back on to the stack.
+	old := *tip
+	switch {
+	case old.prev == nil: // At head position
+		// shift current tip head to the right
+		*tip = old.next
+		// Remove reference to node from chain
+		if old.next != nil { // This is when liquidity hits zero
+			old.next.prev = nil
+		}
+	case old.next == nil: // At tail position
+		// Remove reference to node from chain
+		old.prev.next = nil
+	default:
+		// Reference prior node in chain to next node in chain
+		// bypassing current node
+		old.prev.next = old.next
+		old.next.prev = old.prev
+	}
+	ll.length--
+	return old
+}
+
+// insert inserts at a tip target (can inline)
+func insert(ll *linkedList, tip **node, updt Item, stack *stack) {
```

Comment:
A little hard to explain this one so we will definitely talk about this tomorrow but I'll have a go here anyways. Currently insert has no way of inserting a node at the end since if I'm understanding correctly, we insert a node just before where the tip is pointing. I think insert function should be really generic and should be able to insert a node wherever it needs to be inserted whether it be in the end or the beginning. This will help prevent the user from having to consider which type of insert needs to be used in which scenario.

Reply:
No user should be calling this function at all. It's only used and inlined into the function that requires it. The alignment is handled by the calling function. And it's stripped bare to do only one thing in context with that calling function.